File: libs/counter/ios/CounterBridge.swift
import Foundation
import WebKit

/**
 * Bridge class for counter functionality in iOS
 */
class CounterBridge: NSObject, WKScriptMessageHandler {
    
    // WebView instance
    private weak var webView: WKWebView?
    
    // List of subscribers to counter changes
    private var subscribers = [(Int) -> Void]()
    
    // Current counter value
    private var currentValue: Int = 0
    
    /**
     * Initialize with a WebView
     */
    init(webView: WKWebView) {
        self.webView = webView
        super.init()
        
        // Add script message handler
        webView.configuration.userContentController.add(self, name: "iOSCounterBridge")
        
        // Initialize bridge in WebView
        initialize()
    }
    
    /**
     * Subscribe to counter changes
     * @param subscriber Function to call when counter changes
     * @return Unsubscribe function
     */
    func subscribe(subscriber: @escaping (Int) -> Void) -> () -> Void {
        subscribers.append(subscriber)
        // Immediately notify with current value
        subscriber(currentValue)
        
        // Return unsubscribe function
        return { [weak self] in
            self?.subscribers.removeAll(where: { $0 === subscriber })
        }
    }
    
    /**
     * Notify all subscribers of counter changes
     * @param count New counter value
     */
    private func notifySubscribers(count: Int) {
        currentValue = count
        subscribers.forEach { $0(count) }
    }
    
    /**
     * Handle messages from JavaScript
     */
    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        if message.name == "iOSCounterBridge", let body = message.body as? String {
            if let count = Int(body) {
                notifySubscribers(count)
            }
        }
    }
    
    /**
     * Increment counter
     */
    func increment() {
        sendEvent(type: "counter/increment")
    }
    
    /**
     * Decrement counter
     */
    func decrement() {
        sendEvent(type: "counter/decrement")
    }
    
    /**
     * Reset counter
     */
    func reset() {
        sendEvent(type: "counter/reset")
    }
    
    /**
     * Set counter value
     * @param value New counter value
     */
    func setValue(value: Int) {
        sendEvent(type: "counter/setValue", payload: value)
    }
    
    /**
     * Send event to JavaScript
     */
    private func sendEvent(type: String, payload: Any? = nil) {
        guard let webView = webView else { return }
        
        var eventDict: [String: Any] = ["type": type]
        if let payload = payload {
            eventDict["payload"] = payload
        }
        
        do {
            let eventData = try JSONSerialization.data(withJSONObject: eventDict)
            if let eventJson = String(data: eventData, encoding: .utf8) {
                let escapedJson = eventJson.replacingOccurrences(of: "\"", with: "\\\"")
                let script = "window.sendCounterEvent(\"\(escapedJson)\")"
                webView.evaluateJavaScript(script, completionHandler: nil)
            }
        } catch {
            print("Error creating event JSON: \(error)")
        }
    }
    
    /**
     * Initialize counter bridge in WebView
     */
    private func initialize() {
        guard let webView = webView else { return }
        
        let script = """
        if (window.counterBridgeInitialized) return;
        
        window.counterBridgeInitialized = true;
        
        // Subscribe to counter changes
        if (typeof window.subscribeToCounterBridge === 'function') {
            window.subscribeToCounterBridge(function(count) {
                // Call iOS bridge
                window.webkit.messageHandlers.iOSCounterBridge.postMessage(count.toString());
            });
        } else {
            console.error('subscribeToCounterBridge is not available');
        }
        """
        
        webView.evaluateJavaScript(script, completionHandler: nil)
    }
    
    /**
     * Cleanup when bridge is no longer needed
     */
    func cleanup() {
        webView?.configuration.userContentController.removeScriptMessageHandler(forName: "iOSCounterBridge")
    }
    
    deinit {
        cleanup()
    }
}

File: libs/counter/android/CounterBridge.kt
package com.aha.counter

import android.webkit.JavascriptInterface
import android.webkit.WebView
import org.json.JSONObject

/**
 * Bridge class for counter functionality in Android
 */
class CounterBridge(private val webView: WebView) {
    
    // List of subscribers to counter changes
    private val subscribers = mutableListOf<(Int) -> Unit>()
    
    // Current counter value
    private var currentValue: Int = 0
    
    init {
        // Add JavaScript interface to WebView
        webView.addJavascriptInterface(this, "AndroidCounterBridge")
    }
    
    /**
     * Subscribe to counter changes
     * @param subscriber Function to call when counter changes
     * @return Unsubscribe function
     */
    fun subscribe(subscriber: (Int) -> Unit): () -> Unit {
        subscribers.add(subscriber)
        // Immediately notify with current value
        subscriber(currentValue)
        
        // Return unsubscribe function
        return {
            subscribers.remove(subscriber)
        }
    }
    
    /**
     * Notify all subscribers of counter changes
     * @param count New counter value
     */
    private fun notifySubscribers(count: Int) {
        currentValue = count
        subscribers.forEach { it(count) }
    }
    
    /**
     * Called from JavaScript when counter value changes
     */
    @JavascriptInterface
    fun onCounterChanged(countStr: String) {
        try {
            val count = countStr.toInt()
            notifySubscribers(count)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
    
    /**
     * Increment counter
     */
    fun increment() {
        val event = JSONObject().apply {
            put("type", "counter/increment")
        }
        sendEvent(event.toString())
    }
    
    /**
     * Decrement counter
     */
    fun decrement() {
        val event = JSONObject().apply {
            put("type", "counter/decrement")
        }
        sendEvent(event.toString())
    }
    
    /**
     * Reset counter
     */
    fun reset() {
        val event = JSONObject().apply {
            put("type", "counter/reset")
        }
        sendEvent(event.toString())
    }
    
    /**
     * Set counter value
     * @param value New counter value
     */
    fun setValue(value: Int) {
        val event = JSONObject().apply {
            put("type", "counter/setValue")
            put("payload", value)
        }
        sendEvent(event.toString())
    }
    
    /**
     * Send event to JavaScript
     * @param eventJson Event JSON string
     */
    private fun sendEvent(eventJson: String) {
        val escapedJson = eventJson.replace("\"", "\\\"")
        webView.post {
            webView.evaluateJavascript(
                "window.sendCounterEvent(\"$escapedJson\")",
                null
            )
        }
    }
    
    /**
     * Initialize counter bridge in WebView
     */
    fun initialize() {
        // Add listener for counter changes
        webView.post {
            webView.evaluateJavascript(
                """
                if (window.counterBridgeInitialized) return;
                
                window.counterBridgeInitialized = true;
                
                // Subscribe to counter changes
                if (typeof window.subscribeToCounterBridge === 'function') {
                    window.subscribeToCounterBridge(function(count) {
                        // Call Android bridge
                        if (window.AndroidCounterBridge) {
                            window.AndroidCounterBridge.onCounterChanged(count.toString());
                        }
                    });
                } else {
                    console.error('subscribeToCounterBridge is not available');
                }
                """.trimIndent(),
                null
            )
        }
    }
}